docs/source/class_basics.rst

```diff
@@ -151,6 +151,112 @@ concrete. As with normal overrides, a dynamically typed method can
 implement a statically typed abstract method defined in an abstract
 base class.
 
+.. _metaclasses:
+
+Metaclasses
+***********
+A `metaclass <https://docs.python.org/3/reference/datamodel.html#metaclasses>`_
+is a class that describes the construction and behavior of other classes,
+similarly to how classes describe the construction and behavior of objects.
+The default metaclass is ``type``, but it's possible to use other metaclasses.
+Metaclasses allows one to create "a different kind of class", such as Enums,
+NamedTuples and singletons.
+
+Mypy has some special understanding of ``ABCMeta``, ``EnumMeta``,
+``NamedTupleMeta`` and of course ``GenericMeta``.
+
+Defining a metaclass
+********************
+
+.. code-block:: python
+
+    class M(type):
+
+    class A(metaclass=M):
+        pass
+
+In Python 2, the syntax for defining a metaclass different:
+
+.. code-block:: python
+
+    class A(object):
+        __metaclass__ = M
+
+Mypy also supports using the `six <https://pythonhosted.org/six/#six.with_metaclass>`_
+library to define metaclass in a portable way:
+
+.. code-block:: python
+    
+    import six
+    
+    class A(six.with_metaclass(M)):
+        pass
+
+    @six.add_metaclass(M)
+    class C(object):
+        pass
+
+Metaclass usage example
+***********************
+
+Mypy supports the lookup of attributes in the metaclass:
+
+.. code-block:: python
+
+    from typing import Type, TypeVar, ClassVar
+    T = TypeVar('T')
+
+    class M(type):
+        count: ClassVar[int] = 0
+
+        def make(cls: Type[T]) -> T:
+            M.count += 1
+            return cls()
+    
+    class A(metaclass=M):
+        pass
+    
+    a = A.make()  # make is looked up at M
+    print(A.count)
```

Comment:
Maybe show here the types of ``a`` and ``A.count``.

Reply:
I'm not sure that `reveal_type` is something that the reader wants to see. Perhaps I should add some lookup failures? Or maybe something like `print(A.count + " objects were created")` which demonstrates an actually helpful error message.

Reply:
For the first I mean ``a: A = A.make()`` or using a type comment. For the second, indeed showing an error like ``incompatible arguments for + "int" and "str"`` is better.